### DISCUSSION TITLE
Question about displaying 16-bit RGB images

### DISCUSSION THREAD

**Discussion by Sum-Guy** [arendst/Tasmota#17588] [Q&A] [answered]
I understand that this command (which I have stumbled across in another discussion) should display an image on a connected display screen:

displaytext [x0y0P/filename.rgb:]

I guess you first need to issue the command

displaymode 0

for it to work.

I'm wondering about the exact syntax.  Is the "/" after the P necessary?  Is the ":" at the end of the file name necessary?

Does the file name suffix have to be ".rgb" ?

And I'm wondering about the file itself.  Is this RGB16 format a known/recognized or standardized format?  Would a graphics or image editing software (CorelDraw, Irfanview, etc) know how to create such a file, with presumably a header structure that is known to the tasmota display driver?

I have read about the script editor which has a sub-function to create this type of file, but I can't find the script editor available anywhere for download.  I have tried some on-line image convertors, I found one that will create RGB 565 format file, and have tried to have tasmota display the file, but when I issue the command it is causing tasmota to re-boot.

**Comment by barbudor** [comment]
1. There are complete samples in the Displays page that will leave no doubt on the syntax. Search for `.rgb` in https://tasmota.github.io/docs/Displays/

2. Scripteditor is linked from the Scripting-Language page here https://tasmota.github.io/docs/Scripting-Language/#optional-external-editor

It is indeed missing a clear documentation of the format which is not AFAIK anything standard. You need to reverse engineer the code if you want to find out

**Reply by gemu2015** [reply]
indeed the .rgb file is simply RGB 565 preceded by a "header" => 2 unsigned words x-size and y-size

**Reply by Sum-Guy** [reply]
> 1. There are complete samples in the Displays page that will leave no doubt on the syntax. Search for `.rgb` in https://tasmota.github.io/docs/Displays/

Yes there are samples.  The samples show the presence of / and .rgb and ":"

That does not, to me, indicate anything.  

It does not tell me if "/" is NECESSARY to preceed the file name.

It does not tell me if ".rgb" is NECESSARY to have in the file name.

It does not tell me if ":" is necessary or why it is there.

**Reply by barbudor** [reply]
Just do as the sample
If you want to know more, dig in the code

**Comment by Sum-Guy** [comment, marked as answer]
I figured out how to create a bitmap, which I assume will work for any LCD bitmapped display connected to a tasmota device (specifically I'm working with a 240 x 240 ST7789 TFT display).

I took a sample jpg photo, reduced it to 240 pixels wide, the height ended up being 142 pixels.

I used irfanview to save the reduced image as RGB raw file.  This creates 24 bit file with no header.  R, G and B are 1 byte each, saved in that order.

I wrote some code to create an rgb file that can directly be displayed by tasmota

displaytext [x0y0P/file.rgb:]

As per instructions, first I wrote out the image width and height (each being 2 byte unsigned integer).  

Then read in the pixel color values, each being 8 bits, and bit-shifted them to get 5-bit red, 6 bit green and 5 bit blue values (shifting to the right, keeping the most significant bits).  I then inverted each of them (red = 31 - red, green = 63 - green, blue = 31 - blue).

I then assembled them into a 16-bit unsigned integer = red + green x 32 + blue x 2048 and wrote that out to the rgb file.

When writing out 16 bit integer values, the byte order is low byte then high byte.

**Reply by huyphamnhu** [reply]
@Sum-Guy Could you share your sample output .rgb565 photo and your script to convert RGB raw to tasmota rgb565 please? I tried alot of online converters but none can be used in the `displaytext [x0y0P/file.rgb:]` command.

**Reply by Sum-Guy** [reply]
I see that it was exactly 2 years ago to the day that I worked on this!

I'm attaching a zip file here, it contains the following:

car.jpg - this was the test picture, scaled down to 240 pixels wide
car.raw - this might be an intermediate form of the image, not sure
new.rgb - this is what I think I was uploading to the ESP8266 file system under tasmota for display
display.ini - this is for tasmota?  maybe how I was configuring the display?

I have some intermediate files (red.rgb, blue.rgb, a few others) I don't know if they played a role in the final file.

Edit:  The car.raw file can be opened with irfanview and displayed using this setup:  24 BPP, 240 x 142 image dimensions, color order RGB, Interleaved.

I have these files on the PC where I was directly working with the ESP8266 hardware.  I was using another PC to work with the image file and wrote the code to generate the image files.  That PC is not currently active but I can bring it up and look for the code files, that won't be a problem but note - I was using power basic for this (I know, very old school but it's what I know).  If you still want me to locate and post that code let me know.  But at least try to see if you can display the new.rgb file on your 240 x 240 pixel display to verify it is connected and working.

[RGB_files.zip](https://github.com/user-attachments/files/18321734/RGB_files.zip)

PS:  why doesn't github accept rar files?

**Reply by huyphamnhu** [reply]
Hi, just want to let you know that your `new.rgb` sample is displayed correctly on my setup.
It would be much appreciated if you could find the time to retrieve your conversion code ❤️ 
Btw, `display.ini` is the configuration file to let the new Tasmota Universal Display Driver know how to talk with the display. And I'm testing with the ST7789 too 😃 

PS: to be honest, I have no experience with PowerBASIC. To the point that just a minute ago I don't even know what it is, haha. My goal is somehow replicate your conversion logic to a python script. Hopefully there will be a not-too-big learning curve in the process.

**Reply by huyphamnhu** [reply]
Hello, it's me again.
I wrote a simple python script to do the conversion but the colors are messed up.
Could you take a look at my code please
[Tasmota-RGB565-Converter](https://github.com/huyphamnhu/Tasmota-RGB565-Converter)

**Reply by joba-1** [reply]
As Sum-Guy noted, 16bit byte ordering is important in the file format. Your code might work for you but not for others using cpu‘s with other endianness.

I would use astype('>u2') or astype('<u2') instead of your astype(np.uint16) for the final output.

**Reply by SilentBob83** [reply]
hi guys I am also working now with the tasmota and Display. I got it running to show the exsample.
I would like to load an jpg to the background but i do not get it. Did you manage it ? I use ESP32 so i think i do not need a RBG file ?

**Reply by sfromis** [reply]
AFAICT, for any support of jpeg images, you need a custom build including `#define JPEG_PICTS`, but I'm not sure whether it would work for your use case. I'm usually using PNG images, with good experiences.

**Comment by SilentBob83** [comment]
I Trier that too but it did not work. How Do you call the command? How is the path if i User the on Chip file System? --Gesendet mit der GMX Mail AppAm 16.01.25, 21:16 schrieb sfromis ***@***.***>:

AFAICT, for any support of jpeg images, you need a custom build including #define JPEG_PICTS, but I'm not sure whether it would work for your use case. I'm usually using PNG images, with good experiences.

**Reply by sfromis** [reply]
My test case was using HASPmota, with the following in `pages.jsonl`:
`{"id":10,"obj":"img","x":0,"y":0,"src":"A:/tasmota-logo-rgb.png"}`
(Same path if using LVGL)
![tasmota-logo-rgb](https://github.com/user-attachments/assets/9d9d4bca-cb57-436a-9baa-2ae59c1d1b71)

**Reply by sfromis** [reply]
BTW. If I use the example image from Sum-Guy (from previously in this thread), this command format gets me to see an image `DisplayText [x0y0P/car.rgb:]` (exactly as already mentioned in this thread) but same format fails with PNG.

However, it appears that red and blue was swapped in the image, meaning not "quite right" conversion to the 16-bit RGB565 format.

**Comment by SilentBob83** [comment]
i have tried again the displaytext -> not working.
I would like to test the JSON input. I use IO broker. I guess you sent it via mqtt. can you tell me wich topic name you are using ?

**Reply by sfromis** [reply]
No, I did not use MQTT. As mentioned, the JSON was in the local file `pages.jsonl`, not via a command, and only in the context of having a setup with HASPmota, specifically including a binary with that feature, and an `autoexec.be` like:
```
import haspmota
haspmota.start()
```
The full `pages.jsonl` was:
```
{"page":1}
{"id":10,"obj":"img","x":0,"y":0,"src":"A:/tasmota-logo-rgb.png"}
``` 
With Berry and HASPmota, there are ways of updating images, instead of it all being from a static file, but that's not relevant to simply testing if you can get an image on the screen.

**Comment by SilentBob83** [comment]
Shall i try to paste  this in the berry Script? --Gesendet mit der GMX Mail AppAm 17.01.25, 11:11 schrieb sfromis ***@***.***>:

No, I did not use MQTT. As mentioned, the JSON was in the local file pages.jsonl, not via a command, and only in the context of having a setup with HASPmota, specifically including a binary with that feature, and an autoexec.be like:
import haspmota
haspmota.start()

The full pages.jsonl was:
{"page":1}
{"id":10,"obj":"img","x":0,"y":0,"src":"A:/tasmota-logo-rgb.png"}

With Berry and HASPmota, there are ways of updating images, instead of it all being from a static file, but that's not relevant to simply testing if you can get an image on the screen.

**Reply by sfromis** [reply]
No, I already told you about what files I used, to make a very simple repeatable test case.

However, if you *already* have the `pages.jsonl` file created, the lines for the file `autoexec.be` (which is Berry code executed when restarting) *could* also be entered in the Berry console, assuming that all preconditions are met.

You also need to make sure that the Tasmota status for the display is powered on, like using front page toggle or command.

**Comment by SilentBob83** [comment]
Actually i have display output from the Demo. Where are These infos with the .jsonl file from ? So you say:  Just create this .jsonl and upload ? Then execute for test in the berry console ?--Gesendet mit der GMX Mail AppAm 17.01.25, 11:37 schrieb sfromis ***@***.***>:

No, I already told you about what files I used, to make a very simple repeatable test case.
However, if you already have the pages.jsonl file created, the lines for the file autoexec.be (which is Berry code executed when restarting) could also be entered in the Berry console, assuming that all preconditions are met.
You also need to make sure that the Tasmota status for the display is powered on, like using front page toggle or command.

**Reply by sfromis** [reply]
No. As I just told you, `autoexec.be` is executed when Tasmota boots. For the basic test with two separate files, you need not use the Berry console at all. It is easiest to create the files directly in the Tasmota web UI, Tools / Manage File system.

For details of what you can do with HASPmota (including the `.jsonl` files), read the [docs](https://tasmota.github.io/docs/HASPmota/).

**Comment by SilentBob83** [comment]
Now i tried in the autoexec:

import haspmota
haspmota.start()

in the page.jsonl:
{"page":1}
{"id":10,"obj":"img","x":0,"y":0,"src":"A:/einstein.png"}

UFS:
autoexec.be              2025-01-17T14:17:32       33 
display.ini              2025-01-16T14:45:50         736 
einstein.png             2025-01-16T18:45:33    81551 
einstein.rgb             2025-01-17T14:16:18    57600 
pages.jsonl              2025-01-17T14:17:11       68 

![einstein](https://github.com/user-attachments/assets/c57145b1-f4fb-410b-9ced-2886a73cc69f)

It is not working only black screen. 
Maybe because of lag of Display Ram ?

**Reply by sfromis** [reply]
When doing a restart after having setup two files along with your image, it works for me. Having a board with PSRAM available is **strongly** recommended, I'm not sure if that image might still be "too big" without, but it is not that big. Make sure to carefully read the log output from the boot.

Note that when saving your image here on Github, I got an image size of 24581 bytes, which was what I meant by "not that big", but your directory info says 81551 bytes which likely would be a problem in case of having no PSRAM.

Here's my much smaller image (655 bytes), also works as a test for colors red/green/blue.
![tasmota-logo-rgb](https://github.com/user-attachments/assets/b58fa78d-4f16-44a4-8b18-a5ca7e277023)

**Reply by sfromis** [reply]
Saving the Einstein image after converting to greyscale and 8-bit colors reduced the size to 4523 bytes. 
![einstein-bw](https://github.com/user-attachments/assets/338dfb4d-0e47-414e-a8f3-3cec2b6b237e)

**Comment by SilentBob83** [comment]
even this is not working.... my bad. Again 5 € less for  "YCD fake"  ;-). There are so many variants of it. crazy.
